### PR TITLE
How to share server state between clients?

### PR DESCRIPTION
**Pull request by rkardashov**
In plain Twisted, I use ProtocolFactory instance to share server state between connected clients (represented by Protocol instances). But how can one share some state when using thrift-generated services code?

### CODE REVIEW

**Comment by rkardashov** [comment]
This is just a question, not really a pull request.

**Comment by amilstead** [comment]
@rkardashov I think what might do the trick is: `TBinaryProtocol.TBinaryProtocolFactory` (part of the [thrift python library](https://github.com/apache/thrift/blob/master/lib/py/src/protocol/TBinaryProtocol.py#L238)).

That's the base-provided protocol factory for Twisted-based Thrift Server Apps (used [here](https://github.com/amilstead/unity3d-thrift-twisted/blob/master/python/app/main.py#L26) in the example project in this repo).

I'd say it should generally be safe to build a class that sub-classes it in some way.

Is that what you're looking for?

**Comment by rkardashov** [comment]
Hey @amilstead, thanks for tip!
The question is, how do I access that Protocol (and then its Factory) from the Service code?

Looking at your example code, I see the Service instances share state via database. But what if I just want to store some simple data in memory? I need some object that every Service instance will have access to.

I wonder if Service have some common access to the underlying Twisted classes, like its Protocol. Or, better, is there some Thrift way to share state?

**Comment by rkardashov** [comment]
Sorry, maybe this is a dumb question as I'm just getting to know Thrift and have very little experience in Twisted =)

**Comment by amilstead** [comment]
No not at all; I'm just trying to make sure I understand it well enough.

tl;dr - `Service` has no knowledge of other objects further up the Twisted dispatch chain, but you can always associate the cache with the `Service` itself. You'll want to take precaution if the cache needs to be shared across server processes (i.e. if you're using a pre-fork model or multiple machines running the same app code).

After digging through some of the Twisted Thrift support, it doesn't look like Service objects really have that much access to or control over the underlying `Protocol` or `Transport`, so trying to use those objects in some way to cache values that would then be accessible to the `Service` seems like it isn't really possible.

That being said; it shouldn't preclude you from sharing state on the `Service` object itself, or some other common module (similar to my database access) that interfaces with a networked cache (i.e. memcached or redis) instead of an in-memory cache (e.g. a dictionary).

As far as I can tell, only one instance of the `Service` is created per-server process so if you're only running one process you can throw a dictionary onto the `Service` object on `__init__` and that should persist across multiple calls to the server.

If your application server uses a pre-fork model with the [`create_server`](https://github.com/amilstead/unity3d-thrift-twisted/blob/master/python/app/main.py#L15) function, then each process will have its own instance of `Service` and that dictionary cache will be different for different process runtimes. If that's the case, this is exactly where I'd pull in redis or memcached to help out. 

Hope that is a better answer to what you're asking!

**Comment by rkardashov** [comment]
Great, thanks a lot, it is much clearer now!
I had similar thoughts of using a simple 'static' dict on a Service.

I better start digging the sources myself to get all these answers =)

Regards, RK
